lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -397,7 +397,106 @@ struct ConvolutionToConv2dPattern
                                              adaptor.getOperandConstraints());
 
     rewriter.replaceOp(op, output);
+    return success();
+  }
+};
+//===----------------------------------------------------------------------===//
+// Gather Pattern Matching
+//===----------------------------------------------------------------------===//
 
+struct GatherToEmbeddingConversionPattern
+    : public OpConversionPattern<ttir::GatherOp> {
+  using OpConversionPattern<ttir::GatherOp>::OpConversionPattern;
+
+  LogicalResult checkBasicLegality(ttir::GatherOp op,
```

Comment:
Is this something that operation verifiers can't catch?

Reply:
We should use verifiers for these scenarios. The only place I think it makes sense to do it outside of the verifiers is in dialects that we don't control, like StableHLO - I don't know if there's a better way to do it though.

Reply:
I think the intention was to keep stablehlo to ttir conversion layer as thin as possible. Whilst using verifiers, we would mark valid stablehlo gather ops as invalid since they cannot be lowered into embedding. We followed the TTIR to TTIR Pattern Matching approach. I'm not sure if one approach is better than the other.

Reply:
@sdjordjevicTT / @svuckovicTT, this is during conversion and is checking to see if the `matchAndRewrite` (see below) is possible, i.e. we're able to legally convert the op, which is separate from verification.

Reply:
This one is especially complicated because TTIR gather is a legal op in TTIR, but this rewrite pattern is trying to decomp into embedding op.  This is because stablehlo does not have embedding, it's using the more general op gather.

Reply:
Thanks for the clarification @ddilbazTT and @nsmithtt, I agree with you.